feat(cli) - enhance input UX with double ESC clear

lifefloating opened the pull request:
## TLDR

This PR improves the input experience by implementing two key UX enhancements:
issue: https://github.com/google-gemini/gemini-cli/issues/4450

 **Double ESC to Clear Input**: Press ESC twice to clear the input buffer, with visual feedback showing "Press Escape again to clear."


These changes align gemini-cli with modern CLI tool expectations and significantly improve developer workflow when working with large text inputs, code blocks, or JSON content.

## Dive Deeper

### Key Improvements

**Double ESC Input Clearing**
- First ESC press shows a 1-second prompt: "Press Escape again to clear."
- Second ESC press clears the input buffer and resets completion state
- Proper state management with automatic timeout and cleanup
- Works seamlessly with existing ESC functionality (shell mode exit, suggestion dismissal)


**UI/UX Consistency**
- Unified color scheme: "Press Escape again to clear." uses gray color (consistent with context info)
- Added external clear buffer interface for better separation of concerns

### Technical Implementation

- **App.tsx**: Enhanced keyboard event handling, added clear buffer coordination
- **InputPrompt.tsx**: Implemented double ESC logic, removed conflicting Ctrl+C handling
- Proper React hooks usage with cleanup and dependency management
- Backward compatible with existing functionality

## Reviewer Test Plan

**Testing Double ESC Clear:**
1. Type a long message or paste large code block
2. Press ESC once → should see "Press Escape again to clear." prompt
3. Press ESC again → input should be cleared
4. Test timeout: Press ESC once, wait 1 second → prompt should disappear


**Edge Cases:**
- Test ESC during completions/suggestions
- Test ESC in shell mode
- Test Ctrl+C during streaming responses
- Test ESC after history navigation

**Example Test Scenarios:**
```bash
# Test large input clearing
> This is a very long message that spans multiple lines...
# Press ESC twice → should clear

# Test exit behavior
> # Empty input, press Ctrl+C twice → should exit with stats
```

## Testing Matrix

|          | 🍏  | 🪟  | 🐧  |
| -------- | --- | --- | --- |
| npm run  | ✅  | ❓  |  ❓  |

*Tested on macOS, Windows and Linux testing would be appreciated*

## Linked issues / bugs

This PR addresses common UX pain points mentioned in community feedback:
- Input clearing workflow inefficiency when working with large text
- Inconsistent terminal exit behavior compared to other CLI tools
- Need for better keyboard shortcuts for power users

*- this is a proactive UX improvement*

lifefloating commented:
merge conflict fixed

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -929,7 +975,8 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
                   shellModeActive={shellModeActive}
                   setShellModeActive={setShellModeActive}
-                  focus={isFocused}
+                  onEscapePromptChange={handleEscapePromptChange}
+                  onClearBuffer={handleClearBuffer}
```
please add back focus.

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -878,6 +920,10 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
                     </Text>
+                  ) : showEscapePrompt ? (
+                    <Text color={Colors.Gray}>
+                      Press Escape again to clear.
```
Other UX in gemini CLI says `esc` rather than `Escape`. Please align with that to keep it consistent.

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -465,7 +493,21 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
       }
     } else if (key.ctrl && (input === 'c' || input === 'C')) {
-      handleExit(ctrlCPressedOnce, setCtrlCPressedOnce, ctrlCTimerRef);
+      // Check if input buffer has content
```
why is this an improvement over the previous logic that handled this within InputPrompt?

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -454,7 +465,24 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
     }
 
-    if (key.ctrl && input === 'o') {
+    if (key.escape) {
```
can you move this escape logic into InputPrompt rather than having App.tsx deal with functionality that is focused on input prompt?

lifefloating commented:
hi @jacob314 
Thank you so much for your prompt response. I truly appreciate you taking the time out of your workday to get back to me.

I've read through your message thoroughly, and your insights are very valuable. I plan to revise things later today based on your feedback. I'll also be sure to conduct more extensive testing and incorporate some supplementary test code.

Many thanks!

lifefloating commented:
Hi @jacob314 

Thanks for the review feedback! I've been working on making this more robust as you suggested.

## Updates Made

**Double ESC Clear Implementation:**
- Fixed the reliability issues you mentioned - the double ESC should work consistently now
- Added proper state management with escape counters and timeout handling
- Integrated escape prompt notifications to parent component via callback

**Test Coverage Added:**
- Added 6 comprehensive test cases covering the core functionality
- Tests validate double ESC behavior, state resets, shell mode compatibility, and edge cases
- All 51 tests passing, including the new ones for escape functionality

**Simplified Ctrl+C Approach:**
- Removed the complex Ctrl+C exit changes for now to focus on the core double ESC feature
- This addresses your concern about unclear Ctrl+C improvements

The double ESC clear now works reliably in my testing. First ESC shows a prompt, second ESC clears the buffer. It properly handles timeouts and doesn't interfere with existing ESC functionality (shell mode exit, suggestion dismissal).

Would appreciate another look when you have a chance!

lifefloating commented:
https://github.com/user-attachments/assets/f6bc154f-4d40-481a-9503-d0101fb58d69

jacob314 reviewed (commented):
Thank you for the pull request!
However e already support ctrl-C to clear the input prompt. I'm not sure allowing double escape as well to also clear the input prompt is worth while particularly given this requires 190 lines of somewhat complex code to get right. I also worry that this does not align with vim mode (recently added) where escape has a specific meaning. For now lets hold off on landing this unless there is significant evidence this is functionality the community cares about.

lifefloating commented:
> > Thank you for the pull request! We'll need to make sure this does not activate in vim mode. You may also be able to take inspiration from functionality to support vim mode to make this a bit more custom. After consideration I'm open to approving this once we make sure it doesn't execute while in vim mode.
> 
> Hi @jacob314 I just arrived at the office and am very happy to receive your feedback. Regarding the vim mode, here are the details:
> 
> * When vim mode is enabled: The ESC key will be fully taken over by vim mode for switching modes or clearing pending states. It will not trigger the double-ESC function to clear input.
> * When vim mode is disabled: The double-ESC function to clear input will be used.
> * the handling of vim mode has a higher priority.
> 
> I have just tested it, and indeed, it does not affect the vim mode.
> 
> As for why I find the double-ESC operation useful, it is because in the technical groups I am part of(wechat groups), many people mention that they are accustomed to using double-ESC to clear content. This is because using Ctrl+C twice can sometimes accidentally exit the CLI.
> 
> There is already a check in place for the vim mode. like:
> 
> ```ts
> if (vimHandleInput && vimHandleInput(key)) {
>   return;
> }
> ```
> 
> Or there could be other solutions to make the vim check more reasonable: Add `vimModeEnabled` prop to `InputPrompt`. However, it would add more code, and I understand that the existing code is sufficient.
> 
> and I tested the Vim mode (with logs) yesterday, and it doesn't actually affect this functionality.

 Hi @jacob314   If you're available, could you please take a look at it again for me?  If there are any remaining issues, I'll be happy to make the necessary adjustments as soon as possible.

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -559,7 +565,16 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
+  }, [ctrlCPressedOnce, handleExit]);
+
   useInput((input: string, key: InkKeyType) => {
+    // if ESC key, don't handle it in App, let InputPrompt handle it
```
this comment doesn't seem right. I would remove this block as it is a no-op. whether we return or not has no baring on how InputPrompt does input handling.

jacob314 commented on packages/cli/src/ui/components/InputPrompt.tsx:
```diff
@@ -222,6 +262,7 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
       }
 
       if (key.name === 'escape') {
+        // Handle existing ESC functionality first
```
remove comment. it describes what was changed not what the behavior is.

jacob314 commented on packages/cli/src/ui/components/InputPrompt.tsx:
```diff
@@ -237,13 +278,36 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
+
+        // Handle double ESC for clearing input
+        if (escPressCount === 0) {
+          if (buffer.text.trim() === '') {
```
nit: probably don't want a trim here.
if I accidentally had 20 spaces in the input prompt I would still want to be able to clear with escape.

lifefloating replied:
When the input box is completely empty (buffer.text === ''):

Pressing ESC for the first time → Return directly without any action.

jacob314 commented on packages/cli/src/ui/components/InputPrompt.tsx:
```diff
@@ -402,12 +466,14 @@ export const InputPrompt: React.FC<InputPromptProps> = ({
-          return;
+        } else if (onCtrlCWithEmptyBuffer) {
+          // Buffer is empty, notify parent component to handle exit
+          onCtrlCWithEmptyBuffer();
```
why this change? we shouldn't be too eager to shut down the app. If we want this lets put it in a separate bug plus pr to discuss.

jacob314 commented on packages/cli/src/ui/App.tsx:
```diff
@@ -1013,6 +1033,8 @@ const App = ({ config, settings, startupWarnings = [], version }: AppProps) => {
                   shellModeActive={shellModeActive}
                   setShellModeActive={setShellModeActive}
+                  onEscapePromptChange={handleEscapePromptChange}
+                  onCtrlCWithEmptyBuffer={handleCtrlCWithEmptyBuffer}
```
this seems like some ctrlC handling changes that were left in this pr accidentally

lifefloating replied:
i have removed it

jacob314 reviewed (commented):
Sorry for the slow review. I missed my notifications for this pr.

lifefloating commented:
> Sorry for the slow review. I missed my notifications for this pr.

Hi @jacob314 ,  just review it when you have time. If you don't see it, no big deal. It's just that I merge from main frequently, so the commit history might look a bit messy.

lifefloating commented:
@jacob314  Thank you very much for your review and understanding! 
The code has now been merged and the status is ready.

jacob314 reviewed (approved):
Really appreciate you sticking with this pull request. Looking forward to being able to use escape to clear rather than having to press ctrl-C which also works well but always makes me worried I will kill Gemini CLI.![lgtm](https://i.imgur.com/JgNRalI.png)
